### ISSUE TITLE
poor scaling efficiency of xlnet

### ISSUE THREAD

**Issue by zmsunnyday** [horovod/horovod#1250]
**Environment:**
1. Framework: TensorFlow
2. Framework version:1.13.1
3. Horovod version:0.16.4
4. MPI version:4.0.1
5. CUDA version:10.0
6. NCCL version:2.4.2
7. Python version:3.6.8
8. OS and version:ubuntu16.04
9. GCC version:5.4.0

**Your question:**
I tried to scale xlnet using horovod on two nodes, each with 8 V100 GPUs, while the scaling efficiency is very poor, about 25%. What's more, even ran on one node, 8 gpu cards, the scalling efficiency was also very low, no more than 50%. The connection between two nodes is 25GEth, and each 4 GPUs on the nodes could use p2p according to NCCL INFO. 
When I ran xlnet pretraining on one gpu, the gpu utilization was almost 100%. But when I ran it on 16 gpus , the gpu utilization was about only 50%, which indicate that the communication used  too much time. I also have tried fp16_allreduced, which didn't help at all. 
I would like to know why this could happen and how to solve this problem?

**Comment by abditag2** [comment]
Can you record the timeline during the training and post it here? here are the docs on how to record the timeline: https://github.com/horovod/horovod/blob/master/docs/timeline.rst

**Comment by zmsunnyday** [comment]
> Can you record the timeline during the training and post it here? here are the docs on how to record the timeline: https://github.com/horovod/horovod/blob/master/docs/timeline.rst
Thx @abditag2 for your suggestion. I recorded the timeline, but I have no authority to get it out of our server and post it here. But the timeline help me to diagnose the problem. It showed that the mpi_allgether used up almost all the time during training, the time used by NCCL_allreduce and GPU computing was very short. So the scalling efficiency was low, even thought I tried fp16_allreduce. I think this was caused by the large model size and low computing requirement of xlnet. What should I do to solve this problem without changing the model structue? Since allgather is performed using pure MPI now, could it be performed using NCCL?

**Comment by zmsunnyday** [comment]
Thx @abditag2 for your suggestion. I recorded the timeline, but I have no authority to get it out of our server and post it here. But the timeline help me to diagnose the problem. It showed that the mpi_allgether used up almost all the time during training, the time used by NCCL_allreduce and GPU computing was very short. So the scalling efficiency was low, even thought I tried fp16_allreduce. I think this was caused by the large model size and low computing requirement of xlnet. What should I do to solve this problem without changing the model structue? Since allgather is performed using pure MPI now, could it be performed using NCCL?

**Comment by karakusc** [comment]
@zmsunnyday Can you try passing `sparse_as_dense=True` to `DistributedOptimizer()`?

**Comment by zmsunnyday** [comment]
thx @karakusc .Your suggestion really help me to solve this problem.

**Comment by zmsunnyday** [comment]
Embedding gradients is sparse, so after passing sparse_as_dense=True to DistributedOptimizer(), the scalling efficiency of xlnet on 16 GPUs can be 75%. But I'm still puzzled why horovod use mpi allgather instead of NCCL allgather?

**Comment by karakusc** [comment]
1. Your embedding gradient is a sparse tensor object (`tf.IndexedSlices`), but it is in fact a dense tensor. This is because the embedding matrix is probably used somewhere else other than the initial embedding of words, which generates a dense gradient on it. But since `tf.IndexedSlices` object is *supposed* to be sparse, Horovod still tries to do Allgather on it.

2. Horovod currently does not support NCCL Allgather, but even if it did this is unlikely to result in a big difference in scaling efficiency compared to MPI.

3. If you want to experiment, you can try passing in `-x HOROVOD_HIERARCHICAL_ALLGATHER=1` flag to try hierarchical allgather, though this probably will not have a huge impact either, especially on two nodes.

@alsrgv Do you think it makes sense to print a warning to try `sparse_as_dense=True` if the density of `tf.IndexedSlices` is above a threshold?

**Comment by agemagician** [comment]
@zmsunnyday Could you please share the code for XLNet distributed training?
